### ISSUE TITLE
Graph data structure for MLD

### ISSUE THREAD

**Issue by TheMarex** [Project-OSRM/osrm-backend#3779]
We need a data structure that makes the following operations fast:

1. Scanning all boundary arcs of a node at a certain level
2. Scanning all arcs to nodes of the same cell on level 1
3. Identify when a node is a boundary node on a certain level

For this I propose the following data structure that basically just relies on a `StaticGraph` but enforces a node ordering that is specific to the partition.

1. Sort all nodes by the following criterion:
     1. The highest level for which it is a boundary node (or never)
     2. The cell in which it is contained on that level
     3. Node ID (even better: BFS ID, but we can do that later)

2. Node IDs of boundary nodes at a certain level will be consecutive. That way we can easily implement a fast map: Node ID -> Boundary ID (row/column) for each cell using a vector for each level, without too much memory overhead.

3. We will get a lot of cache efficiency because the search on the level 0 base graph will only be within one part of the adjacency array.

We can create this in the partitioner just ~after we have determined the `MultiLevelPartition` and~ before we will create the `CellStorage`. `CellStorage` will then depend on the graph having a certain ID ordering to facilitate fast lookups.

EDIT: We need to do this before the `MultiLevelPartition` because we store data reference by node ID.

/cc @oxidase

**Comment by TheMarex** [comment]
Will try to sketch this data structure out today.

**Comment by TheMarex** [comment]
Resorting the edges makes us dependent on https://github.com/Project-OSRM/osrm-backend/issues/3737 since the current format of the lookup file assumes edges are always sorted in the way we emit them in the `EdgeBasedGraphFactory`.

**Comment by TheMarex** [comment]
Thinking a little bit more on how to do this, I think it can be done bypassing some of the structural problems here:

1. What we want to re-number are the `EdgeBasedNode` ids. These are referenced in two files after executing `osrm-extract`:
    - `.fileIndex` stores a reference to the `forward`/`backward` node in the search graph.
    - `.ebg` stores edges as `(source, target, data)`

2. We already read the `.ebg` file in `osrm-partition` anyway, we could also renumber every edge and write it out again.

3. Loading the `.fileIndex` file in `osrm-partition` and renumbering it is kind of ugly but works.

4. There should be no downside from doing something like `osrm-extract` then run `osrm-partition` (modified the `.ebg` file) `osrm-contract` (reads the `.ebg` file modied by `osrm-partition`). Best case this will also improve cache locality for CH.

5. Multiple runs of `osrm-partition` are not affected by the renumbered file.

**Comment by TheMarex** [comment]
First progress report: Got the renumbering working up until the `customization` step (the `.ebg_nodes` and `.fileIndex` files still need to be renumbered).
This already enabled me to measure the impact of the new numbering on the customization:

Heat map of the access pattern on the graph during `customization`:

The X-axis shows the node ID space, the Y-axis is the time of execution (axis is flipped, top is start).

**master**
![renumbered_access_master](https://cloud.githubusercontent.com/assets/606968/26365449/2f6274b8-3fd7-11e7-9fd9-2b04a65fe6f5.png)

**renumbered**
![renumbered_access_new](https://cloud.githubusercontent.com/assets/606968/26365451/316ae830-3fd7-11e7-8d6b-59e603548381.png)

As can already be guessed from these images, the renumbering reduces the cache-misses by more then 40%.

That said, this actually seems to decrease performance, which doesn't really make sense. That is the output of `perf`. The number of cycles is higher for this branch. My current theory is that maybe the graph is corrupted, can't really tell before the whole pipeline works.

**master**
```

      57072.444921      task-clock:u (msec)       #    2.871 CPUs utilized          
                 0      context-switches:u        #    0.000 K/sec                  
                 0      cpu-migrations:u          #    0.000 K/sec                  
            11,810      page-faults:u             #    0.207 K/sec                  
   178,176,979,804      cycles:u                  #    3.122 GHz                    
    55,911,885,018      instructions:u            #    0.31  insn per cycle         
     8,453,088,503      branches:u                #  148.112 M/sec                  
       359,245,673      branch-misses:u           #    4.25% of all branches        

      19.875727620 seconds time elapsed
```

**renumbered**

```
      70158.402397      task-clock:u (msec)       #    3.267 CPUs utilized          
                 0      context-switches:u        #    0.000 K/sec                  
                 0      cpu-migrations:u          #    0.000 K/sec                  
            11,691      page-faults:u             #    0.167 K/sec                  
   219,937,312,489      cycles:u                  #    3.135 GHz                    
    54,968,211,339      instructions:u            #    0.25  insn per cycle         
     8,359,567,428      branches:u                #  119.153 M/sec                  
       316,904,267      branch-misses:u           #    3.79% of all branches        

      21.472205157 seconds time elapsed
```

**Comment by TheMarex** [comment]
This was addressed with #4089.
